apis/r/R/SOMAExperimentAxisQuery.R

```diff
@@ -705,6 +705,264 @@ SOMAExperimentAxisQuery <- R6::R6Class(
+        "'obsm_layers' must be a character vector" = is.null(obsm_layers) ||
+          is.character(obsm_layers) ||
+          is_scalar_logical(obsm_layers),
+        "'obsp_layers' must be a character vector" = is.null(obsp_layers) ||
+          is.character(obsp_layers) ||
+          is_scalar_logical(obsp_layers),
+        "'varp_layers' must be a character vector" = is.null(varp_layers) ||
+          is.character(varp_layers) ||
+          is_scalar_logical(varp_layers)
+      )
+      pad_names <- function(x) {
+        stopifnot(
+          is.character(x)
+        )
+        if (is.null(names(x))) {
+          return(stats::setNames(nm = x))
+        }
+        unnamed <- !nzchar(names(x))
+        names(x)[unnamed] <- x[unnamed]
+        return(x)
+      }
+      # Re-implement SingleCellExperiment's internal `.mat2hits` function as
+      # theirs doesn't recognize the difference between `Matrix` and `matrix`
+      # without {Matrix} being attached
+      # Also correct cases where `S4Vectors::SelfHits(from =` and
+      # `S4Vectors::SelfHits(to =` are not integers by explicitly casting to
+      # integer
+      mat_to_hits <- function(mat) {
+        f <- if (inherits(mat, 'Matrix')) {
+          Matrix::which
+        } else {
+          base::which
+        }
+        i <- f(mat != 0, arr.ind = TRUE)
+        return(S4Vectors::SelfHits(
+          from = as.integer(i[, 1L]),
+          to = as.integer(i[, 2L]),
+          nnode = nrow(mat),
+          x = mat[i]
+        ))
+      }
+      # Load in colData
+      obs <- private$.load_df('obs', index = obs_index, columns = obs_column_names)
+      # Load in rowData
+      var <- private$.load_df('var', index = var_index, columns = var_column_names)
+      # Check the layers
+      X_layers <- X_layers %||% self$ms$X$names()
+      X_layers <- pad_names(X_layers)
+      assert_subset(x = X_layers, y = self$ms$X$names(), type = 'X_layer')
+      # Read in the layers
+      layers <- lapply(
+        X = X_layers,
+        FUN = function(layer, var_ids = row.names(var), obs_ids = row.names(obs)) {
+          mat <- private$.as_matrix(self$X(layer), transpose = TRUE)
+          dimnames(mat) <- list(var_ids, obs_ids)
+          return(mat)
+        }
+      )
+      names(layers) <- names(X_layers)
+      # Load in reduced dimensions
+      ms_obsm <- tryCatch(expr = self$ms$obsm$names(), error = null)
+      skip_rd <- isFALSE(obsm_layers) || rlang::is_na(obsm_layers)
+      if (is.null(ms_obsm)) {
+        if (!skip_rd && !is.null(obsm_layers)) {
+          warning(
+            "Reduced dimensions requested but none were found, skipping",
+            call. = FALSE,
+            immediate. = TRUE
+          )
+        }
+        skip_rd <- TRUE
+      }
+      if (skip_rd) {
+        reduced_dims <- list()
+      } else {
+        if (isTRUE(obsm_layers)) {
+          obsm_layers <- NULL
+        }
+        obsm_layers <- obsm_layers %||% stats::setNames(
```

Comment:
I know my feedback is delayed and I should have asked sooner but this is a wonderful opportunity for some simple little helper methods -- it would be easier to review, for sure, but also easier on future maintainers